cloudinit/netinfo.py

```diff
@@ -18,13 +20,74 @@
 
 LOG = logging.getLogger()
 
-
+# Example netdev format:
+# {'eth0': {'hwaddr': '00:16:3e:16:db:54',
+#           'ipv4': [{'bcast': '10.85.130.255',
+#                     'ip': '10.85.130.116',
+#                     'mask': '255.255.255.0',
+#                     'scope': 'global'}],
+#           'ipv6': [{'ip': 'fd42:baa2:3dd:17a:216:3eff:fe16:db54/64',
+#                     'scope6': 'global'},
+#                    {'ip': 'fe80::216:3eff:fe16:db54/64', 'scope6': 'link'}],
+#           'up': True},
+#  'lo': {'hwaddr': '',
+#         'ipv4': [{'bcast': '',
+#                   'ip': '127.0.0.1',
+#                   'mask': '255.0.0.0',
+#                   'scope': 'host'}],
+#         'ipv6': [{'ip': '::1/128', 'scope6': 'host'}],
+#         'up': True}}
 DEFAULT_NETDEV_INFO = {"ipv4": [], "ipv6": [], "hwaddr": "", "up": False}
 
 
+def _netdev_info_iproute_json(ipaddr_json):
+    """Get network device dicts from ip route and ip link info.
+
+    ipaddr_out: Output string from 'ip --json addr' command.
+
+    Returns a dict of device info keyed by network device name containing
+    device configuration values.
+
+    Raises json.JSONDecodeError if json could not be decoded
+    """
+    ipaddr_data = json.loads(ipaddr_json)
+    devs = {}
+
+    for dev in ipaddr_data:
+        flags = dev["flags"]
+        dev_info = {
+            "hwaddr": dev["address"] if dev["link_type"] == "ether" else "",
+            "up": bool("UP" in flags and "LOWER_UP" in flags),
+            "ipv4": [],
+            "ipv6": [],
+        }
+        for addr in dev["addr_info"]:
+            if addr["family"] == "inet":
+                parsed_addr = {
+                    "ip": addr["local"],
+                    "mask": str(
+                        IPv4Network(f'0.0.0.0/{addr["prefixlen"]}').netmask
+                    ),
+                    "bcast": (
+                        addr["broadcast"] if "broadcast" in addr else ""
+                    ),
+                    "scope": addr["scope"],
+                }
+                dev_info["ipv4"].append(parsed_addr)
+            elif addr["family"] == "inet6":
+                parsed_addr = {
+                    "ip": f"{addr['local']}/{addr['prefixlen']}",
```

Comment:
A note that tun0 for PTP `link_type: "none"` will provide `noprefixroute: true` which I think means you shouldn't add the prefix suffix on the local IP in this case as it only applies to the "peer" link.
"""
  If a peer address is specified, the local address cannot
              have a prefix length. The network prefix is associated
              with the peer rather than with the local address
""" 
from https://man7.org/linux/man-pages/man8/ip-address.8.html


Here's the output from --json for my tun0 interface:
```json
  {
    "ifindex": 15,
    "ifname": "tun0",
    "flags": [
      "POINTOPOINT",
      "MULTICAST",
      "NOARP",
      "UP",
      "LOWER_UP"
    ],
    "mtu": 1500,
    "qdisc": "fq_codel",
    "operstate": "UNKNOWN",
    "group": "default",
    "txqlen": 100,
    "link_type": "none",
    "addr_info": [
      {
        "family": "inet",
        "local": "10.172.65.170",
        "prefixlen": 18,
        "broadcast": "10.172.127.255",
        "scope": "global",
        "noprefixroute": true,
        "label": "tun0",
        "valid_life_time": 4294967295,
        "preferred_life_time": 4294967295
      },
      {
        "family": "inet6",
        "local": "2001:67c:1562:8007::aac:41aa",
        "address": "2001:67c:1562:8007::1",
        "prefixlen": 64,
        "scope": "global",
        "noprefixroute": true,
        "valid_life_time": 4294967295,
        "preferred_life_time": 4294967295
      }
    ]
  }
```

Reply:
I think what we want here is "ip": "2001:67c:1562:8007::aac:41aa" instead of "2001:67c:1562:8007::aac:41aa/64"
or better handing of link_type: "none"

Reply:
I don't think that's what `noprefixroute: true` means. That's related to the kernel automatically creating a default route for this address. I think you're right though that if there is a separate `address` and `local`, that the local shouldn't display a mask.

I also wonder if 'link_type' is none if we should include the address at all or perhaps figure out a way to show display it differently in the net device info / route info tables we display from this info.